Sets directory permissions from tar after extraction

bstick12 opened the pull request:
* Allows extraction of tar archives that have read-only directories -
replicating the functionality of `tar --delay-directory-restore`

This is a proposed fix for https://github.com/buildpack/lifecycle/issues/177

jromero commented on archive/tar_test.go:
```diff
@@ -29,6 +29,53 @@ func testTar(t *testing.T, when spec.G, it spec.S) {
+			var err error
+			tmpDir, err = ioutil.TempDir("", "extract-tar-test")
+
+			if err != nil {
```
Typically the assertion below is sufficient.

jromero commented:
Thanks for the contribution! 

Seems like there are some failing tests. They aren't explicitly related to these changes but could there be some cross contamination or unexpected side-effects?

```
--- FAIL: TestRestorer (0.03s)
    --- FAIL: TestRestorer/Restorer (0.03s)
        --- PASS: TestRestorer/Restorer/#Restore/there_is_no_previous_cache/does_nothing (0.00s)
        --- FAIL: TestRestorer/Restorer/#Restore/there_is_a_cache/restores_cached_layers (0.00s)
            restorer_test.go:194: Expected nil: chmod /tmp: operation not permitted
        --- FAIL: TestRestorer/Restorer/#Restore/there_is_a_cache/write_a_.sha_file_for_launch_layers (0.00s)
            restorer_test.go:212: Expected nil: chmod /tmp: operation not permitted
        --- FAIL: TestRestorer/Restorer/#Restore/there_is_a_cache/doesn't_restore_cache_false_layers (0.01s)
            restorer_test.go:236: Expected nil: chmod /tmp: operation not permitted
        --- FAIL: TestRestorer/Restorer/#Restore/there_is_a_cache/doesn't_restore_layers_from_buildpacks_that_aren't_in_the_group (0.00s)
            restorer_test.go:247: Expected nil: chmod /tmp: operation not permitted
        --- FAIL: TestRestorer/Restorer/#Restore/there_is_a_cache/escapes_buildpack_IDs_when_restoring_buildpack_layers (0.00s)
            restorer_test.go:254: Expected nil: chmod /tmp: operation not permitted
        --- SKIP: TestRestorer/Restorer/#Restore/there_is_a_cache/restorer_is_running_as_root/recursively_chowns_the_layers_dir_to_CNB_USER_ID:CNB_GROUP_ID (0.01s)
            restorer_test.go:274: 
```

bstick12 commented:
> Thanks for the contribution!
> 
> Seems like there are some failing tests. They aren't explicitly related to these changes but could there be some cross contamination or unexpected side-effects?
> 
> ```
> --- FAIL: TestRestorer (0.03s)
> ```

My bad not running all the tests. The restorer tests exposed a behaviour I wasn't covering in the added   test case where you should not change the permissions on an existing directory. e.g `/var`. I've modified the behaviour in Untar to take this into account and added a test case.

sclevine commented:
This looks great to me. @ekcasey mind if I merge?
